// sqe/lib/sqe.js
/**
    Licensed to the Mendix Technology B.V. under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.
*/

/*global require, console, module */
var resources = require('./resources/resources'),
    sqlite3 = require('sqlite3').verbose();

function sqe() {
    'use strict';
    var link = null,
        resource = resources('en_EN');
    return {
        link: link,
        open: function (filename, type, callback) {
            this.link = new sqlite3.Database(filename);
        },
        close: function () {
            if (this.link) {
                this.link.close();
            }
        },
        createTable: function (json, callback) {
            var sql = '', action = null, field = null, fieldContent = null;
            sql += 'CREATE TABLE IF NOT EXISTS ' + json.name + '(id INT PRIMARY KEY);\n';
            for (field in json.fields) {
                if (json.fields.hasOwnProperty(field)) {
                    fieldContent = json.fields[field];
                    sql += 'ALTER TABLE ' + json.name + ' ADD COLUMN ' + fieldContent.name + ' ' + fieldContent.ftype + ';\n';
                }
            }
            console.log();
            console.log(resource.CLI_GENERATED_SQL);
            console.log(sql);
            this.link.exec(sql, function (link) {
                callback(json.name);
            });
        },
        showTable: function (json, callback) {
            if (this.link) {
                console.log('');
                console.log(resource.CLI_TABLE_INFORMATION + json.name);
                this.link.each("PRAGMA table_info('" + json.name + "')", function (err, row) {
                    console.log(row.name + ": " + row.type + ((row.pk) ? ' IS PRIMARY KEY' : ''));
                });
                callback();
            }
        }
    };
}

module.exports = sqe;
